src/client.c

```diff
@@ -1012,7 +1076,26 @@ static bool handle_client_work(PgSocket *client, PktHdr *pkt)
 	client->link->ready = false;
 	client->link->idle_tx = false;
 
+	if (ps_action != PS_IGNORE) {
+		switch (pkt->type)
+		{
+			case 'P':
+				return handle_parse_command(client, pkt);
+			case 'B':
+				return handle_bind_command(client, pkt);
+			case 'D':
+				return handle_describe_command(client, pkt);
+		}
+	}
+
 	/* forward the packet */
+	if (is_prepared_statements_enabled(pool) && track_outstanding) {
+		if (!add_outstanding_request(client, pkt->type, RA_FORWARD)) {
+			/* TODO disconnect oom */
```

Comment:
This seems important, if this fails we already processed expect_rfq_count for instance. So we should fail hard here. Another option would be to move this up before `expect_rfq_count`, but we should consider the consequences of that.

Reply:
Frankly speaking I do not  completely understand this fragment of code.
Especially taken in account your last commit [b500a2b](https://github.com/pgbouncer/pgbouncer/pull/845/commits/b500a2b74c10f8b5a12b845c4d4e286b106f9bb6)
You add `track_outstanding = true` for Bind, Describe and Close messages.
But Bind and Describe  messages are handled in the switch above and so for them control even reach this point.
The same is true for Close named statement message.

`track_outstanding = true`  is also set for Query, FuncCall, Sync and Exec messages. 
But I do not understand why it is necessary to forward them as outstanding messages?
Does `sbuf_prepare_send` do the same?

Reply:
> But Bind and Describe messages are handled in the switch above and so for them control even reach this point. The same is true for Close named statement message.

Only if they contain stuff related to named prepared statements. They can also be used for portals or unnamed prepared statements. We cannot only track a few of them, otherwise we might match them to the wrong returning packets. This definitly deserves some code comments though, because indeed it's not obvious.

> `track_outstanding = true` is also set for Query, FuncCall, Sync and Exec messages. But I do not understand why it is necessary to forward them as outstanding messages? 

For Query, FuncCall & Sync: Sync is needed to know the synchronization point in case of errors (see other comment, currently not actually done yet). Because this needs to be tracked we should also to track Query and FuncCall (afaict), to know we're matching the Sync with the correct ReadyForQuery.
For Exec: It seemed nice to have the injected fake packets also be received in order together with the response to Exec, since it's often part of a Parse+Bind+Describe+Exec flow. So if you do that flow twice and the second Parse needs to inject a same message, you would get responses in this order: ParseComplete -> BindComplete -> RowDescription -> **ParseComplete -> CommandComplete** -> BindComplete -> RowDescription -> CommandComplete. Instead of the one that makes sense with respect to the order of the received packets: ParseComplete -> BindComplete -> RowDescription -> **CommandComplete -> ParseComplete** -> BindComplete -> RowDescription -> CommandComplete

> Does `sbuf_prepare_send` do the same?

I'm not entirely sure what you mean with this question. The same as what?